src/common-libs/plan/index.ts

```diff
@@ -0,0 +1,511 @@
+    });
+  }
+};
+
+const getAssociatedPlanVersion = async (
+  models: Database,
+  planTag: InstanceDataOfModel<Database['planTag']>
+) => {
+  const associatedPlanVersion = await models.planVersion.findOne({
+    where: {
+      planId: planTag.planId,
+      createdAt: {
+        [models.Op.LTE]: planTag.createdAt,
+      },
+      latestVersion: true,
+    },
+  });
+  if (!associatedPlanVersion) {
+    throw new Error('Cannot activate planTag if no associated version exist');
+  }
+  return associatedPlanVersion;
+};
+
+export const updateLastPublishedReportingPeriodId = async (
+  models: Database,
+  planTag: InstanceDataOfModel<Database['planTag']>
+) => {
+  const reportingPeriods = await models.planReportingPeriod.find({
+    where: {
+      planId: planTag.planId,
+      measurementsGenerated: true,
+    },
+    orderBy: { column: 'periodNumber', order: 'desc' },
+  });
+
+  if (reportingPeriods.length) {
+    const latestReportingPeriodId = reportingPeriods[0].id;
+    const publishedMeasurements = await models.measurement.find({
+      where: {
+        planReportingPeriodId: latestReportingPeriodId,
+        versionTags: { [models.Op.NOT_IN]: [['{}']] },
+      },
+    });
+    console.log(publishedMeasurements.length);
+    if (publishedMeasurements.length) {
+      const associatedPlanVersion = await getAssociatedPlanVersion(
+        models,
+        planTag
+      );
+      console.log(associatedPlanVersion);
+      await models.planVersion.update({
+        values: {
+          lastPublishedReportingPeriodId: latestReportingPeriodId,
+        },
+        where: {
+          id: associatedPlanVersion.id,
+        },
+      });
+    }
+  }
+};
+
+const versionedModels: any = [
+  { table: 'attachment', relationship: true },
+  { table: 'governingEntity', relationship: true },
+  { table: 'planEntity', relationship: true },
+  { table: 'planLocation' },
+  { table: 'planVersion' },
+  { table: 'planYear' },
+];
+export const updateVersionStates = async (
+  models: Database,
+  planTag: InstanceDataOfModel<Database['planTag']>,
+  publishMeasurements: boolean,
+  planReportingPeriodIds: PlanReportingPeriodId[] | []
+) => {
+  for (const versionedModel of versionedModels) {
+    console.log(versionedModel.table, versionedModel?.relationship);
+    if (versionedModel?.relationship) {
+      await updateRelationshipModelTags(models, versionedModel.table, planTag);
+    } else {
+      await updateEntityVersionModelTags(models, versionedModel.table, planTag);
+    }
+  }
+  if (publishMeasurements) {
+    await updateMeasurementModelTags(
+      models,
+      'measurement',
+      planTag,
+      planReportingPeriodIds
+    );
+  }
+};
+
+const updateRelationshipModelTags = async (
+  models: Database,
+  tableName: 'attachment' | 'governingEntity' | 'planEntity',
+  tag: Pick<
+    InstanceDataOfModel<Database['planTag']>,
+    'createdAt' | 'planId' | 'public' | 'name'
+  >
+) => {
+  const model = models._tables[tableName];
```

Comment:
No need to access "private" field `_tables`, which is exposed for testing purposes.
```suggestion
  const model = models[tableName];
```
Replace other similar usages in this file